frontend/catalyst/jax_tracer.py

```diff
@@ -106,6 +113,59 @@ def _eval_jaxpr(*args):
 }
 
 
+def _promote_jaxpr_types(types: List[List[ShapedArray]]) -> List[ShapedArray]:
+    # TODO: We seem to use AbstractQreg incorrectly, so JAX doesn't recognize it as a valid abstact
+    # value. One need to investigate how to use it correctly and remove the condition [1].
```

Comment:
Interesting! Do you have any more info on how this manifested, is it because it's not a `ShapedArray`?

Reply:
Here is the typical error
```
  File /opt/venv/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py:2206 in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File /opt/venv/lib/python3.10/site-packages/jax/_src/linear_util.py:188 in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File /workspace/modules/catalyst/frontend/catalyst/pennylane_extensions.py:168 in __call__
    jaxpr, shape = trace_quantum_function(self.func, device, args, kwargs)
  File /workspace/modules/catalyst/frontend/catalyst/jax_tracer.py:620 in trace_quantum_function
    qrp_out = trace_quantum_tape(quantum_tape, device, qreg_in, ctx, trace)
  File /workspace/modules/catalyst/frontend/catalyst/jax_tracer.py:374 in trace_quantum_tape
    qrp2 = op.trace_quantum(ctx, device, trace, qrp)
  File /workspace/modules/catalyst/frontend/catalyst/pennylane_extensions.py:948 in trace_quantum
    branch_jaxprs=unify_result_types(jaxprs2),
  File /workspace/modules/catalyst/frontend/catalyst/jax_tracer.py:164 in unify_result_types
    promoted_types = _promote_jaxpr_types([j.out_avals for j in jaxprs])
  File /workspace/modules/catalyst/frontend/catalyst/jax_tracer.py:123 in _promote_jaxpr_types
    results = list(map(partial(reduce, jnp.promote_types), zip(*types)))
  File /opt/venv/lib/python3.10/site-packages/jax/_src/dtypes.py:549 in promote_types
    a_tp = cast(JAXType, a if any(a is t for t in _weak_types) else np.dtype(a))
TypeError: Cannot interpret 'AbstractQreg()' as a data type
```

Reply:
Hmm interesting, thank you!

Reply:
But you see, it didn't actually start, the failure is at some kind of casting, see [here](https://github.com/google/jax/blob/e27af48510b57e5d9f1ae17f256bd67fa9db7941/jax/_src/dtypes.py#L554). Actually I am sure I met similar errors before, that is why I decided to write this Todo.

Reply:
Right, the conversion to `np.dtype(a)` is sure to fail though (since `AbstractQreg` is not a dtype). I think avoiding the promotion of `AbstractQreg` entirely is probably the correct approach 👍

Reply:
Yes, I think so. Still I am not sure about this `weak_types` . We might need to add AbstractQreg to this list. I am updating the Todo with a note about it.

Reply:
`weak_types` are meant for Python `int`, `float`, etc. The reason is that the bitwidth is not fixed on those types, and you probably don't want a `float32` to be casted to `float64` we combined with a Python `float`. That's why the Python types are regarded as weak, so they don't take precedence over `dtypes`.
You can find the details of the JAX promotion rules here: https://jax.readthedocs.io/en/latest/type_promotion.html